resources/shaders/denoising_reflections.hlsl

```diff
@@ -0,0 +1,424 @@
+RWTexture2D<float> out_history_texture : register(u1);
+RWTexture2D<float2> out_moments_texture : register(u2);
+
+SamplerState point_sampler : register(s0);
+SamplerState linear_sampler : register(s0);
+
+cbuffer CameraProperties : register(b0)
+{
+	float4x4 view;
+	float4x4 projection;
+	float4x4 inv_projection;
+	float4x4 inv_view;
+	float4x4 prev_projection;
+	float4x4 prev_view;
+	uint is_hybrid;
+	uint is_path_tracer;
+	uint is_ao;
+	uint has_shadows;
+	uint has_reflections;
+	float3 padding1;
+};
+
+cbuffer DenoiserSettings : register(b1)
+{
+	float color_integration_alpha;
+	float moments_integration_alpha;
+	float variance_clipping_sigma;
+	float roughness_reprojection_threshold;
+	int max_history_samples;
+	float n_phi;
+	float z_phi;
+	float l_phi;
+};
+
+cbuffer WaveletPass : register(b2)
+{
+	float wavelet_size;
+}
+
+float3 OctToDir(uint octo)
+{
+	float2 e = float2( f16tof32(octo & 0xFFFF), f16tof32((octo>>16) & 0xFFFF) ); 
+	float3 v = float3(e, 1.0 - abs(e.x) - abs(e.y));
+	if (v.z < 0.0)
+		v.xy = (1.0 - abs(v.yx)) * (step(0.0, v.xy)*2.0 - (float2)(1.0));
+	return normalize(v);
+}
+
+float Luminance(float3 color)
+{
+	return (color.r + color.r + color.b + color.g + color.g + color.g) / 6.0;
+}
+
+void FetchNormalAndLinearZ(in float2 ipos, out float3 norm, out float2 zLinear)
+{
+	norm = normal_metallic_texture[ipos].xyz;
+	zLinear = linear_depth_texture[ipos].xy;
+}
+
+float NormalDistanceCos(float3 n1, float3 n2, float power)
+{
+	return pow(max(0.0, dot(n1, n2)), power);
+}
+
+float ComputeWeightNoLuminance(float depth_center, float depth_p, float phi_depth, float3 normal_center, float3 normal_p)
+{
+	const float w_normal    = NormalDistanceCos(normal_center, normal_p, 128.f);
+	const float w_z         = abs(depth_center - depth_p) / phi_depth;
+
+	return exp(-max(w_z, 0.0)) * w_normal;
+}
+
+bool IsReprojectionValid(float2 coord, float z, float z_prev, float fwidth_z, float3 normal, float3 normal_prev, float fwidth_normal)
+{
+	int2 screen_size = int2(0, 0);
+	output_texture.GetDimensions(screen_size.x, screen_size.y);
+
+	bool ret = (coord.x >= 0 && coord.x < screen_size.x && coord.y >= 0 && coord.y < screen_size.y);
+
+	ret = ret && ((abs(z_prev - z) / (fwidth_z + 1e-4)) < 8.0);
+
+	ret = ret && ((distance(normal, normal_prev) / (fwidth_normal + 1e-2)) < 16.0);
+
+	return ret;
+}
+
+bool LoadPrevData(float2 screen_coord, inout float2 found_pos, out float4 prev_direct, out float2 prev_moments, out float history_length)
+{
+	float2 screen_size = float2(0.f, 0.f);
+	output_texture.GetDimensions(screen_size.x, screen_size.y);
+
+	float2 uv = screen_coord / screen_size;
+
+	float4 motion = motion_texture.SampleLevel(point_sampler, uv, 0);
+	float2 q_uv = uv - motion.xy;
+
+	float2 prev_coords = q_uv * screen_size;
+
+	const float roughness = albedo_roughness_texture[screen_coord].w;
+
+	float4 depth = linear_depth_texture[prev_coords];
+	float3 normal = OctToDir(asuint(depth.w));
+	prev_direct = float4(0, 0, 0, 0);
```

Comment:
In shaders you should explicitly state the types to avoid conversion, 0 will be seen as an int. 0.0 will be seen as double. Ill add a comment on every line where this is applicable.